README.md
# proba_sdk_flutter

Mobile framework for Proba platform.

## Installation

1. Add this to your package's pubspec.yaml file:
```
    dependencies:
      proba_sdk_flutter:
```

2. Now in your Dart code, you can use:
```
    import 'package:proba_sdk_flutter/proba_sdk_flutter.dart';
```

## Usage


### Initialization:

```
    await Proba.initialize(
      appId: "${YOUR_APP_ID}",
      sdkToken: "${YOUR_SDK_TOKEN}",
      deviceId: "${YOUR_DEVICE_ID}" // optional, UUID generated by default
      appsFlyerId: "${YOUR_APPS_FLYER_UID}" // optional
      amplitudeUserId: "${YOUR_AMPLITUDE_USER_ID}" // optional, if Amplitude integration is needed
      defaults: {"${TEST_1_KEY}": "${TEST_1_DEFAULT_VALUE}", "${TEST_2_KEY}": "${TEST_2_DEFAULT_VALUE}"},
      deviceProperties: {"installedAt": "2021-05-20T09:55:05.000+03:00"}, // optional, additional information about device
    );
```

After initialization all functionality will be available via singleton:

```
    Proba.instance()
```

### How to fetch known test values that associated with your device?

```
    Proba.instance()!.loadExperiments((List<String> experimentsKeys) {
        // Handle loaded experiments
    });
```

### How to get the value for a specific test?

```
    final value = Proba.instance()!.experiment("${TEST_KEY}");
```

In case of problems with no internet connection or another, the values obtained in the previous session will be used, or if they are missing, the default values specified during initialization will be used.

### How to get user tests for analytics?

```
    final value = Proba.instance()!.experiments;
    
    // or if you need additional details for experiments
    final value = Proba.instance()!.experimentsWithDetails;

```

### How to debug?

Before debug make sure that debug-mode for your App is turned-on on [settings page](https://platform.proba.ai/ab/settings)

  ![](https://imgproxy.proba.ai/9ACImnEbmsO822dynjTjcC_B8aXzbbpPQsOgop2PlBs//aHR0cHM6Ly9hcHBib29zdGVyLWNsb3VkLnMzLmV1LWNlbnRyYWwtMS5hbWF6b25hd3MuY29tLzk0N2M5NzdmLTAwY2EtNDA1Yi04OGQ4LTAzOTM4ZjY4OTAzYi5wbmc.png)

In debug mode you can see all actual tests and check how the user will see each option of the test.
To show the debug bottom sheet you just need to turn it on in your personal cabinet and call

```
    Proba.instance()!.showDebugLayer(
      context: context,
      valuesChangedCallback: (List<String> changedExperimentsKeys) {
        // Handle changed experiments
      },
    );
```

You can enable showing debug bottom sheet by performing shake motion on your device:

```
    Proba.instance()!.enableDebugOnShake(
      context: context,
      valuesChangedCallback: (List<String> changedExperimentsKeys) {
        // Handle changed experiments
      },
    );
```

or disable this behavior:

```
    Proba.instance()!.disableDebugOnShake();
```


CHANGELOG.md
## [1.0.0] - 20/09/2021

## [2.0.0] - 24/09/2021

* Null safety support added.

example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:proba_sdk_flutter/proba_sdk_flutter.dart';
import 'package:pretty_json/pretty_json.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Proba Flutter SDK Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage() : super();

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  static const _experimentsDefaults = {
    "flutter_test": "green",
    "test_experiment": "value_1"
  };
  static const _deviceProperties = {
    "installedAt": "2021-05-20T09:55:05.000+03:00",
  };
  bool _debugOnShake = false;
  bool _showProgress = false;

  Future<void> _initializeSdk() async {
    setState(() => _showProgress = true);
    await Proba.initialize(
      appId: "16897",
      sdkToken: "E44A1C2E762B41A691494FAB045993DF",
      defaults: _experimentsDefaults,
      deviceProperties: _deviceProperties,
    );
    if (!mounted) return;
    setState(() => _showProgress = false);
  }

  void _fetchExperiments() {
    if (_showProgress) return;
    setState(() => _showProgress = true);
    Proba.instance()?.loadExperiments((_) {
      if (!mounted) return;
      setState(() => _showProgress = false);
    });
  }

  bool _sdkInitialized() => Proba.instance() != null;

  String _experimentsStr(Map? experiments) {
    if (experiments == null) return '—';
    return "\n${prettyJson(experiments, indent: 2)}";
  }

  bool _isDebugAllowed() => Proba.instance()?.isDebugAllowed ?? false;

  void _showDebug(BuildContext context) {
    Proba.instance()?.showDebugLayer(
      context: context,
      valuesChangedCallback: (_) => setState(() {}),
    );
  }

  void _toggleDebugOnShake(BuildContext context, bool enabled) {
    if (enabled) {
      Proba.instance()?.enableDebugOnShake(
        context: context,
        valuesChangedCallback: (_) => setState(() {}),
      );
    } else {
      Proba.instance()?.disableDebugOnShake();
    }
    setState(() {
      _debugOnShake = enabled;
    });
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    return Scaffold(
      appBar: AppBar(
        title: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Flexible(child: Text('Proba SDK Test')),
            if (_showProgress)
              CircularProgressIndicator(
                  valueColor: AlwaysStoppedAnimation<Color>(Colors.white)),
          ],
        ),
      ),
      body: Container(
        padding: const EdgeInsets.all(16),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.start,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: <Widget>[
            Padding(
              padding: const EdgeInsets.only(bottom: 8.0),
              child: Text(
                "Defaults:\n${_experimentsDefaults.toString()}",
                style: theme.textTheme.subtitle1,
              ),
            ),
            Padding(
              padding: const EdgeInsets.only(bottom: 8.0),
              child: Text(
                "deviceProperties:\n${_deviceProperties.toString()}",
                style: theme.textTheme.subtitle1,
              ),
            ),
            Wrap(
              alignment: WrapAlignment.spaceEvenly,
              direction: Axis.horizontal,
              children: [
                ElevatedButton(
                  onPressed: _sdkInitialized() ? null : _initializeSdk,
                  child: Text('Initialize SDK'),
                ),
                ElevatedButton(
                  onPressed: _sdkInitialized() ? _fetchExperiments : null,
                  child: Text('Fetch Experiments'),
                ),
                ElevatedButton(
                  onPressed:
                      _isDebugAllowed() ? () => _showDebug(context) : null,
                  child: Text('Show debug'),
                ),
                IgnorePointer(
                  ignoring: !_isDebugAllowed(),
                  child: Row(
                    children: [
                      Checkbox(
                        value: _debugOnShake,
                        onChanged: (value) =>
                            _toggleDebugOnShake(context, value!),
                      ),
                      Text(
                        'Use shake',
                        style: theme.textTheme.button!.copyWith(
                            color: _isDebugAllowed()
                                ? Colors.black87
                                : Colors.black26),
                      ),
                    ],
                  ),
                )
              ],
            ),
            Padding(
              padding: const EdgeInsets.only(top: 16.0),
              child: Text(
                  "Current experiments values: ${_experimentsStr(Proba.instance()?.experiments)}"),
            ),
            Padding(
              padding: const EdgeInsets.only(top: 16.0),
              child: Text(
                  "Current detailed experiments values: ${_experimentsStr(Proba.instance()?.experimentsWithDetails)}"),
            ),
          ],
        ),
      ),
    );
  }
}
